src/SqlPersistence/Config/SqlDialect_MsSqlServer.cs

```diff
@@ -34,7 +34,18 @@ internal override void SetParameterValue(DbParameter parameter, object value)
                 if (value is ArraySegment<char> charSegment)
                 {
                     parameter.Value = charSegment.Array;
-                    parameter.Size = charSegment.Count;
+
+                    // Set to 4000 or -1 to improve query execution plan reuse
+                    // Must be when exceeding 4000 characters for nvarchar(max)  https://stackoverflow.com/a/973269/199551
+                    parameter.Size = charSegment.Count > 4000 ? -1 : 4000;
```

Comment:
We know what are the sizes of the columns in the tables the persisted is using. Would it be possible to change the signature of the method so that the callers would be forced to pass that information?